beginner_source/profiler_tutorial.py

```diff
@@ -0,0 +1,205 @@
+#
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Improve memory performance
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+# Note that the most expensive operations - in terms of memory and time -
+# are at ``forward (11)`` or where we get the mask indices. Let’s try to
+# tackle the memory consumption first. We can see that the ``.to()``
+# operation consumes 953.67 Mb. ``mask`` is initialized with a
+# ``torch.double`` datatype. Can we reduce the memory footprint by casting
+# it to ``torch.float``?
+#
+
+model = CustomLinear(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.float).cuda()
+
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Improve time performance
+# ^^^^^^^^^^^^^^^^^^^^^^^^^
+# That worked, the memory footprint has halved - but while the time
+# consumed has reduced a bit (previously 283ms), it’s still too high at
+# 194ms. Turns out copying a matrix from CUDA to CPU is pretty expensive!
+# The ``aten::copy_`` operator in ``forward (11)`` copies ``mask`` to CPU
+# so that it can use the NumPy ``argwhere`` function. We could eliminate
+# the need to copy into a NumPy array if we use a ``torch`` function
+# ``nonzero()`` here instead.
+#
+
+class CustomLinearLite(torch.nn.Linear):
+    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
+        super().__init__(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        # with profiler.record_function("LINEAR PASS"):
+            out = super().forward(input)
+
+        # with profiler.record_function("INDEX SCORE"):
+            threshold = out.sum(axis=1).mean()
+            hi_idx = (mask > threshold).nonzero(as_tuple=True)
+
+            return out, hi_idx
+
+
+model = CustomLinearLite(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.float).cuda()
+
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Replacing ``np.argwhere`` with ``torch.nonzero`` reduced the total
+# runtime to about 36ms (from 195ms) by eliminating the need to copy to
+# CPU and instead running on the GPU.
+#
+
+
+######################################################################
+# Caveat
+# ^^^^^^^^^^^^^^
+#
+# You might notice introducing ``torch.nonzero`` in our module has shown a
+# new memory footprint of 2.59 Gb that wasn’t in the stacktrace when we
+# used ``np.argwhere``. This is because Profiler tracks only operations in
```

Comment:
let me check what's going on here;

also, more precisely, profiler tracks PyTorch tensor memory consumption, and does not track numpy tensors, but we plan to have a better memory tracking

Reply:
Gotcha, I'll update the note to reflect this.